database/tbcd/level/level.go

```diff
@@ -419,13 +504,367 @@ func (l *ldb) BlockHeaderGenesisInsert(ctx context.Context, wbh *wire.BlockHeade
+// blocks is not allowed, but this is correct behavior as removing chunks of
+// headers in the reverse order they were originally added will ensure that
+// a header being removed only has a maximum of one child (which must be included
+// in the headers passed to this function).
+//
+// For example given a chain:
+//
+//	_______/-[2a]-[3a]-[4a]
+//
+// [ G]-[ 1]-[2b]-[3b]-[4b]-[5b]
+//
+//	____________\-[3c]-[4c]-[5c]-[6c]
+//
+// Where the tip is [6c], the next removal could for example be:
+//
+//	[3a]-[4a]
+//	[3b]-[4b]-[5b]
+//	[5c]-[6c] (and pass in tipAfterRemoval=[5b])
+//
+// But the next removal could not for example be:
+//
+//	[2a]-[3a] // Leaves [4a] dangling
+//	[2b]-[3b]-[4b]-[5b] // Leaves "c" fork dangling
+//
+// The upstream user of a TBC instance in External Header mode is expected
+// to always remove chunks of headers in the opposite order they were
+// originally added. While this is not checked explicitly, failure to do so
+// can result in these types of dangling chain scenarios. In the above example,
+// block [2b] must have been added at or before the time of adding [3b] and [3c].
+//
+// It could have either been:
+// Update #1: ADD [2b]-[3b]-[4b]-[5b]
+// Update #2: ADD [3c]-[4c]-[5c]-[6c]
+// OR
+// Update #1: ADD [2b]-[3c]-[4c]-[5c]-[6c]
+// Update #2: ADD [3b]-[4b]-[5b]
+// (Or some similar order where some of the higher b/c blocks were added back
+// and forth between the chains or split into multiple smaller updates.)
+//
+// Assuming the upstream caller needs to remove the entire b and c chains:
+//
+// If it was the first order, then we would expect upstream caller to first
+// remove [3c]-...-[6c] (undo update #2), and then remove [2b]-...-[5b] (undo
+// update #1), which would never leave a chain dangling.
+//
+// Similarly, if it was the second order, then we would expect upstream caller
+// to first remove [3b]-...-[5b] (undo update #2), and then remove [2b]-...-[6c]
+// (undo update #1) which would also never leave a chain dangling.
+//
+// If the upstream caller removed [2b]-...-[5b] first, then they did not
+// remove headers in the same order that they added them, because it would
+// have been impossible to originally add [2b] after adding [3c].
+//
+// Calling this function with the incorrect tipAfterRemoval WILL FAIL as that
+// indicates incorrect upstream behavior.
+//
+// If any of the above requirements are not true, this function will return
+// an error. If this function returns an error, no changes have been made to
+// the underlying database state as all validity checks are done before db
+// modifications are applied.
+//
+// If an upstreamCursor is provided, it is updated atomically in the database
+// along with the state transition of removing the block headers.
+func (l *ldb) BlockHeadersRemove(ctx context.Context, bhs *wire.MsgHeaders, tipAfterRemoval *wire.BlockHeader, upstreamStateId *[32]byte) (tbcd.RemoveType, *tbcd.BlockHeader, error) {
```

Comment:
This is very complicated and seems to be doing a bit too much. Not sure how to fix it yet but jotting down initial findings.

The big question is: do we have to remove blockheaders? and why?